File: src/main/java/com/gsdd/course/app/config/RequestLogger.java
package com.gsdd.course.app.config;

import com.gsdd.course.app.CourseApplication;
import java.util.concurrent.TimeUnit;
import lombok.extern.slf4j.Slf4j;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.springframework.stereotype.Component;

@Slf4j
@Aspect
@Component
public class RequestLogger {

  @Around("execution(* " + CourseApplication.BASE_PACKAGE + "controller.CourseController.*(..))")
  public Object around(ProceedingJoinPoint joinPoint) throws Throwable {
    Object o;
    String nameJp = joinPoint.getSignature().getName();
    long startTime = System.nanoTime();
    log.info("Before: {}", nameJp);
    o = joinPoint.proceed();
    log.info("After: {}", nameJp);
    long timeTaken = System.nanoTime() - startTime;
    log.info(
        "request {} took {} ms",
        nameJp,
        TimeUnit.MILLISECONDS.convert(timeTaken, TimeUnit.NANOSECONDS));
    return o;
  }
}

File: src/main/kotlin/com/gsdd/course/app/config/RequestLogger.kt
package com.gsdd.course.app.config

import com.gsdd.course.app.CourseApplication
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.springframework.stereotype.Component
import java.util.concurrent.TimeUnit
import org.slf4j.Logger
import org.slf4j.LoggerFactory

inline fun <reified T> T.logger(): Logger {
    return LoggerFactory.getLogger(T::class.java)
}

@Aspect
@Component
class RequestLogger {
    companion object {
        val log  = logger()
    }

    @Around("execution(* com.gsdd.course.app.controller.CourseController.*(..))")
    @Throws(Throwable::class)
    fun around(joinPoint: ProceedingJoinPoint): Any? {
        val nameJp = joinPoint.signature.name
        val startTime = System.nanoTime()
        log.info("Before: {}", nameJp)
        val o = joinPoint.proceed()
        log.info("After: {}", nameJp)
        val timeTaken = System.nanoTime() - startTime
        log.info(
            "request {} took {} ms",
            nameJp,
            TimeUnit.MILLISECONDS.convert(timeTaken, TimeUnit.NANOSECONDS)
        )
        return o
    }
}
